pyrolite/mineral/normative.py

```diff
@@ -301,308 +302,661 @@ def CIPW_norm(df, Fe_correction=None, adjust_all=False):
-    leucite = np.where(leu_flt_2, norm["orthoclase"], leucite)
-    orthoclase = np.where(
-        leu_flt_1, norm["orthoclase"] - (D_SiO2 / 2), norm["orthoclase"]
-    )
-    orthoclase = np.where(leu_flt_2, 0, norm["orthoclase"])
-    norm["leucite"] = leucite
-    norm["orthoclase"] = orthoclase
-
-    D_SiO2 = np.where(leu_flt_2, D_SiO2 - (2 * norm["orthoclase"]), 0)
-
-    # Dicalcium silicate/wollastonite ###################
-    cs_flt_1 = (D_SiO2 > 0) & (D_SiO2 < norm["wollastonite"] / 2)  # if deficit < woll
-    cs_flt_2 = (D_SiO2 > 0) & (D_SiO2 >= norm["wollastonite"] / 2)  # if deficit > woll
-    cs = np.where(cs_flt_1, D_SiO2, 0)
-    norm["cs"] = np.where(cs_flt_2, norm["wollastonite"], cs)
-    wollastonite = np.where(
-        cs_flt_1, norm["wollastonite"] - (2 * D_SiO2), norm["wollastonite"]
-    )
-    wollastonite = np.where(cs_flt_2, 0, wollastonite)
-    norm["wollastonite"] = wollastonite
 
-    D_SiO2 = np.where(cs_flt_2, D_SiO2 - (norm["wollastonite"] / 2), 0)
+    # Normative Zircon
+    df['Z'] = df['n_ZrO2']
+    df['Y'] = df['Z']
 
-    # Dicalcium silicate/olivine ###################
-    oli_flt_1 = (D_SiO2 > 0) & (D_SiO2 < norm["diopside"])  # if deficit < diopside
-    oli_flt_2 = (D_SiO2 > 0) & (D_SiO2 >= norm["diopside"])  # if deficit > diopside
-    olivine = np.where(oli_flt_1, norm["olivine"] + (D_SiO2 / 2), norm["olivine"])
-    olivine = np.where(oli_flt_2, norm["olivine"] + (norm["diopside"] / 2), olivine)
-    norm["olivine"] = olivine
+    # Normative apatite
 
-    cs = np.where(oli_flt_1, norm["cs"] + (D_SiO2 / 2), norm["cs"])
-    norm["cs"] = np.where(oli_flt_2, norm["cs"] + (norm["diopside"] / 2), cs)
+    df['Ap'] = np.where(
+        df['n_CaO'] >= (3 + 1 / 3) * df['n_P2O5'], df['n_P2O5'], df['n_CaO'] / (3 + 1 / 3)
+    ).T
 
-    diopside = np.where(oli_flt_1, norm["diopside"] - D_SiO2, norm["diopside"])
-    diopside = np.where(oli_flt_2, 0, diopside)
-    norm["diopside"] = diopside
+    df['n_CaO_'] = np.where(
+        df['n_CaO'] >= (3 + 1 / 3) * df['n_P2O5'], df['n_CaO'] - (3 + 1 / 3) * df['Ap'], 0
+    ).T
 
-    D_SiO2 = np.where(oli_flt_2, D_SiO2 - norm["diopside"], 0)
+    df['n_P2O5_'] = np.where(
+        df['n_CaO'] < (3 + 1 / 3) * df['n_P2O5'], df['n_P2O5'] - df['Ap'], 0).T
 
-    # Kaliophilite/leucite ###################
-    kal_flt_1 = (D_SiO2 > 0) & (norm["leucite"] >= (D_SiO2 / 2))  # if deficit < leucite
-    kal_flt_2 = (D_SiO2 > 0) & (norm["leucite"] < (D_SiO2 / 2))  # if deficit > leucite
-    norm["kaliophilite"] = np.where(kal_flt_1, D_SiO2 / 2, 0)
-    norm["kaliophilite"] = np.where(kal_flt_2, norm["leucite"], norm["kaliophilite"])
+    df['n_CaO'] = df['n_CaO_']
+    df['n_P2O5'] = df['n_P2O5_']
 
-    leucite = np.where(kal_flt_1, norm["leucite"] - (D_SiO2 / 2), norm["leucite"])
-    leucite = np.where(kal_flt_2, 0, leucite)
-    norm["leucite"] = leucite
-    # Subdivide hy & di into Mg & Fe??
+    df['FREE_P2O5'] = df['n_P2O5']
```

Comment:
For some things like this which you aren't necessarily returning later, you don't need to add them to the dataframe - consider just having variables to hose these series.

Reply:
for now i have left as is